### PR TITLE
feat: lock application/pipeline/environment as code

### PR DESCRIPTION
**Pull request by sguiheux**
1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

File: sdk/pipeline.go [hunk 1]
```diff
@@ -6,16 +6,17 @@ import (
...
+	Name           string      `json:"name" cli:"name,key"`
+	Description    string      `json:"description" cli:"description"`
+	ProjectKey     string      `json:"projectKey"`
+	ProjectID      int64       `json:"-"`
+	Stages         []Stage     `json:"stages"`
+	Parameter      []Parameter `json:"parameters,omitempty"`
+	Usage          *Usage      `json:"usage,omitempty"`
+	Permission     int         `json:"permission"`
+	LastModified   int64       `json:"last_modified" cli:"modified"`
+	FromRepository string      `json:"from_repository" cli:"modified"`
```
Comment (yesnault) [inline sdk/pipeline.go]:
```suggestion
	FromRepository string      `json:"from_repository" cli:"from_repository"`
```

File: engine/api/application/application_parser.go [hunk 2]
```diff
@@ -34,15 +40,20 @@ func ParseAndImport(db gorp.SqlExecutor, cache cache.Store, proj *sdk.Project, e
 	}
 
 	//If the application exist and we don't want to force, raise an error
-	if oldApp != nil && !force {
+	if oldApp != nil && !opts.Force {
 		return nil, nil, sdk.ErrApplicationExist
 	}
 
+	if oldApp.FromRepository != "" && (oldApp.FromRepository == "" || opts.FromRepository != oldApp.FromRepository) {
```
Comment (fsamin) [inline engine/api/application/application_parser.go]:
```suggestion
	if oldApp.FromRepository != "" && opts.FromRepository != oldApp.FromRepository {
```

File: engine/api/pipeline/pipeline_importer.go [hunk 3]
```diff
@@ -26,11 +26,16 @@ func ImportUpdate(db gorp.SqlExecutor, proj *sdk.Project, pip *sdk.Pipeline, msg
 		return sdk.WrapError(err, "Unable to load pipeline %s %s", proj.Key, pip.Name)
 	}
 
+	if oldPipeline.FromRepository != "" && (oldPipeline.FromRepository == "" || pip.FromRepository != oldPipeline.FromRepository) {
```
Comment (bnjjj) [inline engine/api/pipeline/pipeline_importer.go]: to update

**Comment by ovh-cds** [comment]
CDS Report ut-engine#8863.0 ✘  
* Stage 1
  * Engine Test with DB ✘

**Comment by ovh-cds** [comment]
CDS Report it#8875.0 ✘  
* w-it
  * w-it ✘

File: engine/api/application.go [hunk 4]
```diff
@@ -402,6 +402,10 @@ func (api *API) updateApplicationHandler() service.Handler {
 			return sdk.WrapError(errloadbyname, "updateApplicationHandler> Cannot load application %s", applicationName)
 		}
 
+		if app.FromRepository != "" {
+			return sdk.ErrForbidden
```
Comment (richardlt) [inline engine/api/application.go]: Wrap the error.

File: engine/api/application.go [hunk 5]
```diff
@@ -466,6 +470,9 @@ func (api *API) postApplicationMetadataHandler() service.Handler {
 		if err != nil {
 			return sdk.WrapError(err, "postApplicationMetadataHandler")
 		}
+		if app.FromRepository != "" {
+			return sdk.ErrForbidden
```
Comment (richardlt) [inline engine/api/application.go]: Wrap the error.

File: engine/api/application_deployment.go [hunk 6]
```diff
@@ -73,6 +73,9 @@ func (api *API) postApplicationDeploymentStrategyConfigHandler() service.Handler
 		if err != nil {
 			return sdk.WrapError(err, "unable to load application")
 		}
+		if app.FromRepository != "" {
+			return sdk.ErrForbidden
```
Comment (richardlt) [inline engine/api/application_deployment.go]: Wrap the error.

File: engine/api/application_deployment.go [hunk 7]
```diff
@@ -144,6 +147,9 @@ func (api *API) deleteApplicationDeploymentStrategyConfigHandler() service.Handl
...
```
Comment (richardlt) [inline engine/api/application_deployment.go]: Wrap the error.

File: engine/api/application_key.go [hunk 8]
```diff
@@ -48,6 +48,9 @@ func (api *API) deleteKeyInApplicationHandler() service.Handler {
 		if errA != nil {
 			return sdk.WrapError(errA, "deleteKeyInApplicationHandler> Cannot load application")
 		}
+		if app.FromRepository != "" {
+			return sdk.ErrForbidden
```
Comment (richardlt) [inline engine/api/application_key.go]: Wrap the error.

File: engine/api/application_key.go [hunk 9]
```diff
@@ -101,6 +104,10 @@ func (api *API) addKeyInApplicationHandler() service.Handler {
 		}
 		newKey.ApplicationID = app.ID
 
+		if app.FromRepository != "" {
+			return sdk.ErrForbidden
```
Comment (richardlt) [inline engine/api/application_key.go]: Wrap the error.

File: engine/api/application_variable.go [hunk 10]
```diff
@@ -101,6 +101,9 @@ func (api *API) deleteVariableFromApplicationHandler() service.Handler {
 		if err != nil {
 			return sdk.WrapError(err, "Cannot load application: %s", appName)
 		}
+		if app.FromRepository != "" {
+			return sdk.ErrForbidden
```
Comment (richardlt) [inline engine/api/application_variable.go]: Wrap the error.

File: engine/api/application_variable.go [hunk 11]
```diff
@@ -153,6 +156,9 @@ func (api *API) updateVariableInApplicationHandler() service.Handler {
...
```
Comment (richardlt) [inline engine/api/application_variable.go]: Wrap the error.

File: engine/api/application_variable.go [hunk 12]
```diff
@@ -204,6 +210,9 @@ func (api *API) addVariableInApplicationHandler() service.Handler {
 		if err != nil {
 			return sdk.WrapError(err, "Cannot load application %s ", appName)
 		}
+		if app.FromRepository != "" {
+			return sdk.ErrForbidden
```
Comment (richardlt) [inline engine/api/application_variable.go]: Wrap the error.

File: engine/api/environment.go [hunk 13]
```diff
@@ -209,6 +209,10 @@ func (api *API) updateEnvironmentHandler() service.Handler {
 			return sdk.WrapError(errEnv, "updateEnvironmentHandler> Cannot load environment %s", environmentName)
 		}
 
+		if env.FromRepository != "" {
+			return sdk.ErrForbidden
```
Comment (richardlt) [inline engine/api/environment.go]: Wrap the error.

File: engine/api/environment/environment_parser.go [hunk 14]
```diff
@@ -32,16 +37,20 @@ func ParseAndImport(db gorp.SqlExecutor, cache cache.Store, proj *sdk.Project, e
...
-	if oldEnv != nil && !force {
+	if oldEnv != nil && !opts.Force {
 		return nil, nil, sdk.ErrEnvironmentExist
 	}
 	if oldEnv != nil {
 		exist = true
 	}
 
+	if oldEnv != nil && oldEnv.FromRepository != "" && opts.FromRepository != oldEnv.FromRepository {
+		return nil, nil, sdk.ErrForbidden
```
Comment (richardlt) [inline engine/api/environment/environment_parser.go]: Wrap the error.

File: engine/api/environment_key.go [hunk 15]
```diff
@@ -44,6 +44,9 @@ func (api *API) deleteKeyInEnvironmentHandler() service.Handler {
 		if errE != nil {
 			return sdk.WrapError(errE, "deleteKeyInEnvironmentHandler> Cannot load environment")
 		}
+		if env.FromRepository != "" {
+			return sdk.ErrForbidden
```
Comment (richardlt) [inline engine/api/environment_key.go]: Wrap the error.

File: engine/api/environment_key.go [hunk 16]
```diff
@@ -93,6 +96,10 @@ func (api *API) addKeyInEnvironmentHandler() service.Handler {
 		}
 		newKey.EnvironmentID = env.ID
 
+		if env.FromRepository != "" {
+			return sdk.ErrForbidden
```
Comment (richardlt) [inline engine/api/environment_key.go]: Wrap the error.

File: engine/api/environment_variable.go [hunk 17]
```diff
@@ -86,6 +86,9 @@ func (api *API) deleteVariableFromEnvironmentHandler() service.Handler {
 		if errEnv != nil {
 			return sdk.WrapError(errEnv, "deleteVariableFromEnvironmentHandler: Cannot load environment %s", envName)
 		}
+		if env.FromRepository != "" {
+			return sdk.ErrForbidden
```
Comment (richardlt) [inline engine/api/environment_variable.go]: Wrap the error.

File: engine/api/environment_variable.go [hunk 18]
```diff
@@ -143,6 +146,9 @@ func (api *API) updateVariableInEnvironmentHandler() service.Handler {
 		if errEnv != nil {
 			return sdk.WrapError(errEnv, "updateVariableInEnvironmentHandler: cannot load environment %s", envName)
 		}
+		if env.FromRepository != "" {
+			return sdk.ErrForbidden
```
Comment (richardlt) [inline engine/api/environment_variable.go]: Wrap the error.

File: engine/api/environment_variable.go [hunk 19]
```diff
@@ -200,6 +206,9 @@ func (api *API) addVariableInEnvironmentHandler() service.Handler {
 		if errEnv != nil {
 			return sdk.WrapError(errEnv, "addVariableInEnvironmentHandler: Cannot load environment %s", envName)
 		}
+		if env.FromRepository != "" {
+			return sdk.ErrForbidden
```
Comment (richardlt) [inline engine/api/environment_variable.go]: Wrap the error.

File: engine/api/pipeline.go [hunk 20]
```diff
@@ -43,6 +43,10 @@ func (api *API) updatePipelineHandler() service.Handler {
 			return sdk.WrapError(err, "cannot load pipeline %s", name)
 		}
 
+		if pipelineDB.FromRepository != "" {
+			return sdk.ErrForbidden
```
Comment (richardlt) [inline engine/api/pipeline.go]: Wrap the error.

File: engine/api/pipeline.go [hunk 21]
```diff
@@ -86,6 +90,14 @@ func (api *API) postPipelineRollbackHandler() service.Handler {
 		db := api.mustDB()
 		u := deprecatedGetUser(ctx)
 
+		pipDB, err := pipeline.LoadPipeline(db, key, name, false)
+		if err != nil {
+			return sdk.WrapError(err, "cannot load pipeline")
+		}
+		if pipDB.FromRepository != "" {
+			return sdk.ErrForbidden
```
Comment (richardlt) [inline engine/api/pipeline.go]: Wrap the error.

File: engine/api/pipeline_job.go [hunk 22]
```diff
@@ -43,6 +43,10 @@ func (api *API) addJobToStageHandler() service.Handler {
 		if errl != nil {
 			return sdk.WrapError(sdk.ErrPipelineNotFound, "addJobToStageHandler> Cannot load pipeline %s for project %s: %s", pipelineName, projectKey, errl)
 		}
+		if pip.FromRepository != "" {
+			return sdk.ErrForbidden
```
Comment (richardlt) [inline engine/api/pipeline_job.go]: Wrap the error.

File: engine/api/pipeline_job.go [hunk 23]
```diff
@@ -152,6 +156,9 @@ func (api *API) updateJobHandler() service.Handler {
 		if err != nil {
 			return sdk.WrapError(err, "cannot load pipeline %s", pipName)
 		}
+		if pipelineData.FromRepository != "" {
+			return sdk.ErrForbidden
```
Comment (richardlt) [inline engine/api/pipeline_job.go]: Wrap the error.

File: engine/api/pipeline_job.go [hunk 24]
```diff
@@ -245,6 +252,9 @@ func (api *API) deleteJobHandler() service.Handler {
 		if errl != nil {
 			return sdk.WrapError(errl, "deleteJobHandler>Cannot load pipeline %s", pipName)
 		}
+		if pipelineData.FromRepository != "" {
+			return sdk.ErrForbidden
```
Comment (richardlt) [inline engine/api/pipeline_job.go]: Wrap the error.

File: engine/api/pipeline_parameter.go [hunk 25]
```diff
@@ -43,6 +43,9 @@ func (api *API) deleteParameterFromPipelineHandler() service.Handler {
 		if err != nil {
 			return sdk.WrapError(err, "deleteParameterFromPipelineHandler: Cannot load %s", pipelineName)
 		}
+		if p.FromRepository != "" {
+			return sdk.ErrForbidden
```
Comment (richardlt) [inline engine/api/pipeline_parameter.go]: Wrap the error.

File: engine/api/pipeline_parameter.go [hunk 26]
```diff
@@ -84,6 +87,9 @@ func (api *API) updateParameterInPipelineHandler() service.Handler {
 		if err != nil {
 			return sdk.WrapError(err, "updateParameterInPipelineHandler: Cannot load %s", pipelineName)
 		}
+		if p.FromRepository != "" {
+			return sdk.ErrForbidden
```
Comment (richardlt) [inline engine/api/pipeline_parameter.go]: Wrap the error.

File: engine/api/pipeline_parameter.go [hunk 27]
```diff
@@ -134,6 +140,9 @@ func (api *API) addParameterInPipelineHandler() service.Handler {
 		if err != nil {
 			return sdk.WrapError(err, "addParameterInPipelineHandler: Cannot load %s", pipelineName)
 		}
+		if p.FromRepository != "" {
+			return sdk.ErrForbidden
```
Comment (richardlt) [inline engine/api/pipeline_parameter.go]: Wrap the error.

File: engine/api/stage.go [hunk 28]
```diff
@@ -30,6 +30,9 @@ func (api *API) addStageHandler() service.Handler {
 		if err != nil {
 			return sdk.WrapError(err, "addStageHandler")
 		}
+		if pipelineData.FromRepository != "" {
+			return sdk.ErrForbidden
```
Comment (richardlt) [inline engine/api/stage.go]: Wrap the error.

File: engine/api/stage.go [hunk 29]
```diff
@@ -116,6 +119,9 @@ func (api *API) moveStageHandler() service.Handler {
 		if err != nil {
 			return err
 		}
+		if pipelineData.FromRepository != "" {
+			return sdk.ErrForbidden
```
Comment (richardlt) [inline engine/api/stage.go]: Wrap the error.

File: engine/api/stage.go [hunk 30]
```diff
@@ -190,6 +196,9 @@ func (api *API) updateStageHandler() service.Handler {
...
```
Comment (richardlt) [inline engine/api/stage.go]: Wrap the error.

File: engine/api/stage.go [hunk 31]
```diff
@@ -243,6 +252,9 @@ func (api *API) deleteStageHandler() service.Handler {
 		if err != nil {
 			return sdk.WrapError(err, "Cannot load pipeline %s", pipelineKey)
 		}
+		if pipelineData.FromRepository != "" {
+			return sdk.ErrForbidden
```
Comment (richardlt) [inline engine/api/stage.go]: Wrap the error.

File: engine/api/workflow.go [hunk 32]
```diff
@@ -407,6 +407,10 @@ func (api *API) putWorkflowHandler() service.Handler {
 			return sdk.WrapError(errW, "putWorkflowHandler> Cannot load Workflow %s", key)
 		}
 
+		if oldW.FromRepository != "" {
+			return sdk.ErrForbidden
```
Comment (richardlt) [inline engine/api/workflow.go]: Wrap the error.

**Comment by ovh-cds** [comment]
CDS Report build-engine-cli-tools#8923.0 ✘  
* Build minimal
  * cdsctl linux ✔ 
  * check SQL Files number ✔ 
  * Engine Build linux amd64 ✘  
  * tools/debpacker ✔ 
  * tools/os-ansible-inventory ✔ 
  * Worker linux amd64 / 386 ✔

**Comment by ovh-cds** [comment]
CDS Report ut-engine#8923.0 ✘  
* Stage 1
  * Engine Test with DB ✘

**Comment by ovh-cds** [comment]
CDS Report build-engine-cli-tools#8926.0 ✘  
* Build minimal
  * cdsctl linux ✘  
  * check SQL Files number ✔ 
  * Engine Build linux amd64 ✔ 
  * tools/debpacker ✔ 
  * tools/os-ansible-inventory ✔ 
  * Worker linux amd64 / 386 ✔

**Comment by ovh-cds** [comment]
CDS Report it#8926.2 ✘  
* w-it
  * w-it ✘

**Pull request merged**